### PR TITLE
[HOPS-1541] provenance

### PR DESCRIPTION
**Pull request by o-alex**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added and passed (for bug fixes / features)
- [x] HOPS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit
- [x] The commit message has the following format: [HOPS-XXX] message

* **Post a link to the associated JIRA issue**
https://hopshadoop.atlassian.net/browse/HOPS-1541

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature - provenance

* **What is the new behavior (if this is a feature change)?**
Log filesystem operations into a log tailed by epipe

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
This JIRA has changes to the database (add tables)

* **Other information**:

### CODE REVIEW

File: src/main/java/io/hops/metadata/hdfs/entity/FileProvXAttrBufferEntry.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  public String getValue() {
+    return value;
+  }
+
+  public byte[] getValueBytes() {
+    return getXAttrBytes(value);
+  }
+
+  public final static byte[] getXAttrBytes(String val) {
```
Comment (maismail) [inline src/main/java/io/hops/metadata/hdfs/entity/FileProvXAttrBufferEntry.java]: Same as below.
return StoredXAttr. getXAttrBytes(val)

Comment (o-alex) [reply src/main/java/io/hops/metadata/hdfs/entity/FileProvXAttrBufferEntry.java]: fixed.
I removed the static methods and used your methods where i was using my own before.

File: src/main/java/io/hops/metadata/hdfs/entity/FileProvXAttrBufferEntry.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    if (val == null) {
+      return null;
+    }
+    return val.getBytes(Charsets.UTF_8);
+  }
+
+  public final static String getXAttrString(byte[] val) {
+    if (val == null) {
+      return new String("");
```
[the remark this answers is not in the document]
Comment (o-alex) [reply src/main/java/io/hops/metadata/hdfs/entity/FileProvXAttrBufferEntry.java]: fixed.
I removed the static methods and used your methods where i was using my own before.

File: src/main/java/io/hops/metadata/hdfs/entity/FileProvXAttrBufferEntry.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+import com.google.common.base.Charsets;
+
+public final class FileProvXAttrBufferEntry {
+
+  private final long iNodeId;
+  private final byte namespace;
+  private final String name;
+  private final int iNodeLogicalTime;
+  private final String value;
```
Comment (maismail) [inline src/main/java/io/hops/metadata/hdfs/entity/FileProvXAttrBufferEntry.java]: You should use byte[] for value and then provide the method to convert to string when needed. Check the StoredXAttr, to be consistent with it

Comment (o-alex) [reply src/main/java/io/hops/metadata/hdfs/entity/FileProvXAttrBufferEntry.java]: ok, switched to a PrimaryKey and byte[] value, mirroring StoredXAttr.

**Pull request merged**